renderlib/CCamera.cpp

```diff
@@ -12,21 +87,188 @@ CCamera::getFrame() const
   return LinearSpace3f(m_U, m_V, -m_N);
 }
 
+// Credit: https://stackoverflow.com/a/32410473/2373034
+// Returns the intersection line of the 2 planes
+Ray
+GetPlanesIntersection(const Plane& p1, const Plane& p2)
+{
+  glm::vec3 p3Normal = glm::cross(p1.normal, p2.normal);
+  float det = glm::length2(p3Normal);
+
+  return Ray(((glm::cross(p3Normal, p2.normal) * p1.d) + (glm::cross(p1.normal, p3Normal) * p2.d)) / det, p3Normal);
+}
+
+// this implementation comes from an old unity3d wiki page.
+// Two non-parallel lines which may or may not touch each other have a point on each line which are closest
+// to each other. This function finds those two points. If the lines are not parallel, the function
+// outputs true, otherwise false.
+bool
+FindClosestPointsOnTwoLines(Ray line1, Ray line2, glm::vec3& closestPointLine1, glm::vec3& closestPointLine2)
+{
+  closestPointLine1 = glm::vec3(0);
+  closestPointLine2 = glm::vec3(0);
+
+  glm::vec3 line1Direction = line1.direction;
+  glm::vec3 line2Direction = line2.direction;
+
+  float a = glm::dot(line1Direction, line1Direction);
+  float b = glm::dot(line1Direction, line2Direction);
+  float e = glm::dot(line2Direction, line2Direction);
+
+  float d = a * e - b * b;
+  if (d == 0.0f) {
+    return false;
+  }
+  // if d==0 then lines are parallel and this is a big fail.
+
+  glm::vec3 r = line1.origin - line2.origin;
+  float c = glm::dot(line1Direction, r);
+  float f = glm::dot(line2Direction, r);
+
+  float s = (b * f - c * e) / d;
+  float t = (a * f - c * b) / d;
+
+  closestPointLine1 = line1.origin + line1Direction * s;
+  closestPointLine2 = line2.origin + line2Direction * t;
+  return true;
+}
+
+bool
+IsOutermostPointInDirection(int pointIndex,
+                            const glm::vec3& direction,
+                            const std::array<glm::vec3, 8>& boundingBoxPoints)
+{
+  glm::vec3 point = boundingBoxPoints[pointIndex];
+  for (int i = 0; i < 8; i++) {
+    if (i != pointIndex && glm::dot(direction, boundingBoxPoints[i] - point) > 0) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+glm::vec3
+CalculateCameraPosition(const CCamera& camera, const CBoundingBox& sceneBBox, float padding = 0.0f)
```

Comment:
What camera position is this calculating?

Reply:
will fix.  This function is kind of unnecessary anyway as it's just directly called by the other.